libs/langchain/langchain/vectorstores/tiledb.py

```diff
@@ -0,0 +1,771 @@
+
+
+def dependable_tiledb_import() -> Any:
+    """Import tiledb-vector-search if available, otherwise raise error."""
+    try:
+        import tiledb as tiledb
+        import tiledb.vector_search as tiledb_vs
+    except ImportError:
+        raise ValueError(
+            "Could not import tiledb-vector-search python package. "
+            "Please install it with `conda install -c tiledb tiledb-vector-search` "
+            "or `pip install tiledb-vector-search`"
+        )
+    return tiledb_vs, tiledb
+
+
+def get_vector_index_uri(uri: str) -> str:
+    return f"{uri}/{VECTOR_INDEX_NAME}"
+
+
+def get_documents_array_uri(uri: str) -> str:
+    return f"{uri}/{DOCUMENTS_ARRAY_NAME}"
+
+
+class TileDB(VectorStore):
+    """Wrapper around TileDB vector database.
+
+    To use, you should have the ``tiledb-vector-search`` python package installed.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import TileDB
+            embeddings = OpenAIEmbeddings()
+            db = TileDB(embeddings, index_uri, metric)
+
+    """
+
+    def __init__(
+        self,
+        embedding: Embeddings,
+        index_uri: str,
+        metric: str,
+        vector_index_uri: str = "",
+        docs_array_uri: str = "",
+        config: Optional[Mapping[str, Any]] = None,
+        timestamp: Any = None,
+        **kwargs: Any,
+    ):
+        """Initialize with necessary components."""
+        self.embedding = embedding
+        self.embedding_function = embedding.embed_query
+        self.index_uri = index_uri
+        self.metric = metric
+        self.config = config
+        self.vector_index_uri = (
+            vector_index_uri
+            if vector_index_uri != ""
+            else get_vector_index_uri(self.index_uri)
+        )
+        self.docs_array_uri = (
+            docs_array_uri
+            if docs_array_uri != ""
+            else get_documents_array_uri(self.index_uri)
+        )
+
+        tiledb_vs, tiledb = dependable_tiledb_import()
+        group = tiledb.Group(self.vector_index_uri, "r")
+        self.index_type = group.meta.get("index_type")
+        group.close()
+        self.timestamp = timestamp
+        if self.index_type == "FLAT":
+            self.vector_index = tiledb_vs.flat_index.FlatIndex(
+                uri=self.vector_index_uri,
+                config=self.config,
+                timestamp=self.timestamp,
+                **kwargs,
+            )
+        elif self.index_type == "IVF_FLAT":
+            self.vector_index = tiledb_vs.ivf_flat_index.IVFFlatIndex(
+                uri=self.vector_index_uri,
+                config=self.config,
+                timestamp=self.timestamp,
+                **kwargs,
+            )
+
+    @property
+    def embeddings(self) -> Optional[Embeddings]:
+        # TODO: Accept embedding object directly
+        return None
+
+    def process_index_results(
+        self,
+        ids: List[int],
+        scores: List[float],
+        k: int = 4,
```

Comment:
can we add `*` to all non-inherited methods to separate args and kwargs

Reply:
Applied.

Reply:
ah sorry should clarify, we can (unfortunately) only do this for methods that aren't already defined on the  base `VectorStore`. so we can't do it for similarity_search_by_vector, similarity_search, max_marginal_relevance_search_by_vector, max_marginal_relevance_search,  add_texts, from_texts

Reply:
Thanks for the clarification! 
Applied.